src/scene/materials/standard-material-options-builder.js

```diff
@@ -0,0 +1,317 @@
+Object.assign(pc, function () {
+    var StandardMaterialOptionsBuilder = function () {
+    };
+
+    StandardMaterialOptionsBuilder.prototype.update = function (device, scene, stdMat, objDefs, staticLightList, pass, sortedLights, prefilteredCubeMap128) {
+
+        var options = {
+            pass: pass,
+            alphaTest: stdMat.alphaTest > 0,
+            forceFragmentPrecision: stdMat.forceFragmentPrecision,
+            chunks: stdMat.chunks,
+            blendType: stdMat.blendType,
+            forceUv1: stdMat.forceUv1
+        };
+
+        if (objDefs) {
+            options.screenSpace = (objDefs & pc.SHADERDEF_SCREENSPACE) !== 0;
+            options.skin = (objDefs & pc.SHADERDEF_SKIN) !== 0;
+            options.useInstancing = (objDefs & pc.SHADERDEF_INSTANCING) !== 0;
+        }
+
+        var minimalOptions = pass > pc.SHADER_FORWARDHDR && pass <= pc.SHADER_PICK;
+
+        if (minimalOptions) {
+            var optionsMin = this._generateMinOptions(stdMat);
+            Object.assign(options, optionsMin);
```

Comment:
I'm not suggesting you change this, but I do notice that this function creates JS objects, then calls Object.assign and then throws away the original objects. This will generate more garbage and probably be a bit slower than the original implementation. It may be that the amount of memory we're talking about here is very minor and the performance difference is negligible. But it's good to be aware of these things... 😄

Reply:
Yeah, I understand that. Ideally, these all (or at least some) options should be cached or pre-build.